### PR TITLE
[SPARK-8202] [PYSPARK] fix infinite loop during external sort in PySpark

### PR DESCRIPTION
**Pull request by davies**
The batch size during external sort will grow up to max 10000, then shrink down to zero, causing infinite loop.
Given the assumption that the items usually have similar size, so we don't need to adjust the batch size after first spill.

cc @JoshRosen @rxin @angelini

### CODE REVIEW

File: python/pyspark/shuffle.py [hunk 1]
```diff
@@ -512,9 +512,6 @@ def load(f):
                     f.close()
                 chunks.append(load(open(path, 'rb')))
                 current_chunk = []
-                gc.collect()
```
Comment (angelini) [inline python/pyspark/shuffle.py]: Why is the call to gc.collect being removed?

**Comment by airhorns** [comment]
@rxin out of curiosity do you remember why you wanted to shrink the batch in https://github.com/apache/spark/commit/3134c3fe495862b7687b5aa00d3344d09cd5e08e ?  Also, @pwendell heads up, this regression is present in `1.4.0-rc4`!

**Comment by kevincox** [comment]
Also why keep the batch size once you know you are going to spill to disk.  All that does is force you to draw from the iterator in batches.  Once you know how big your chunk size should be you can set `batch` to `len(current_chunk)` (possibly `*0.8` or something) so that you can do it in a single call since you already know you will be spilling to disk.

File: python/pyspark/shuffle.py (hunk 1 above)
Comment (davies) [inline python/pyspark/shuffle.py]: We won't change `limit` to `_next_limit()` (which call get_used_memory()). This line here was to get better number about how much memory was used, is not needed anymore.

**Comment by davies** [comment]
@airhorns That change is make by me (my mistake), to speed up the unit test of external sorting. Because we adjust the batch size during external aggregation (the size of objects vary from key to key), just borrowed that logic here without notice that `elif not chunks:` below.

@pwendell  Could we not block 1.4 release by this one? 1.4 release is delayed much and many people are waiting for it. This issue could be workaround by increasing the number of partitions during sorting (this could also improve performance by avoid spilling). 

@kevincox Is the current approach something similar what you had suggested?

**Comment by rxin** [comment]
@kevinco it was not me -- I merely fixed a merge conflict and somehow git showed me as the commit author.

**Comment by kevincox** [comment]
@davies No, it appears that you just changed the original memory limit.  I am saying once you figure out how large the chunk can be you should set that as the batch size, rather then continuing to allocate in chunks.

**Comment by rxin** [comment]
Is there a way we can add a test for this?

**Comment by JoshRosen** [comment]
Maybe we can borrow some of the testing strategies from Tungsten's external sorters, where we're able to mock out the shuffle memory manager and are able to manually trigger spills.

**Comment by davies** [comment]
@JoshRosen I had updated the tests to reproduce the failure.

**Comment by davies** [comment]
@JoshRosen  The last commit had passed the tests, last two tests failed in other places.

**Comment by davies** [comment]
@JoshRosen Is there something wrong with the NewPullRequestBuilder? the failed two run was triggered by my bookmarklet.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: python/pyspark/tests.py [hunk 2]
```diff
@@ -179,9 +179,12 @@ def test_in_memory_sort(self):
                          list(sorter.sorted(l, key=lambda x: -x, reverse=True)))
 
     def test_external_sort(self):
+        class CustomizedSorter(ExternalSorter):
+            def _next_limit(self):
```
Comment (JoshRosen) [inline python/pyspark/tests.py]: I think that we should add a comment here to explain why we're mocking out this part of the code; it doesn't seem self-evident to me and I'm worried that it's going to confuse future readers of this code.

Also, do you think that it's worth adding a separate test case for this path and keeping the old test?  There might be some duplication of the code which does assertions over metrics, but we possibly can factor it out into a shared method.

Comment (JoshRosen) [inline python/pyspark/tests.py]: It seems like the intent here is to mock `get_used_memory()`; I'm not super familiar with Python mocking frameworks, but if this was Java then I imagine that `get_used_memory()` would be a method of some context / environment object that we could mock.

**Comment by JoshRosen** [comment]
It looks like this is essentially reverting the changes made in the test speedup patch, so at least from that perspective it seems like a sensible fix to me.

Just to make sure that I 100% understand the logic / considerations here, it seems like the method that we're modifying is trying to buffer as many records as possible in memory until we hit a memory limit, then sort the in-memory records and spill them to disk.  It's expensive to figure out how much memory we're actually using, which is why we only check our memory usage every `batch` items.  If our objects are so big that `batch` of them are going to blow way past our memory limit, then we used to try to shrink `batch` so that we inspect the sizes more frequently.  The old logic seemed wrong, though, since it would _always_ shrink the batch size even if we managed to buffer many chunks before spilling.  Instead, if we did want to adjust our batch size then I think it would make more sense to explicitly deal with the case where `num_chunks = 1` because a single chunk is exceeding our memory limit.

@kevincox, to touch on your comment, I think that the reason that we don't just set `batch` to `len(current_chunk)` after spilling is that our objects might be of different sizes, so completely eliminating batching / periodic size checking greatly increases the chance that we'll exceed our memory limits well before we've checked our memory consumption.  @davies, could you confirm that this is the right rationale?

File: python/pyspark/tests.py (hunk 2 above)
Comment (davies) [inline python/pyspark/tests.py]: Yes, without the mock, it will take a long time to reach memory limit (slowing down tests).

**Comment by JoshRosen** [comment]
After this patch, it looks like `_next_limit` is only called once per `ExternalSorter.sorted` call.  Since `_next_limit` is only called from there, I wonder whether we should just remove that method and inline its code at the call site in order to make things a bit easier to read.

This makes me wonder, though: is it actually safe to not allow the memory limit to raise after spilling like it did before?  Here's the comment in `_next_limit`:

```python
    def _next_limit(self):
        """
        Return the next memory limit. If the memory is not released
        after spilling, it will dump the data only when the used memory
        starts to increase.
        """
        return max(self.memory_limit, get_used_memory() * 1.05)
```

If we no longer call `_next_limit()` after spilling, what will happen in cases where spilling somehow prevented memory from being freed?  Does this mean that we'll spill every `batch` items and pay a huge cost during merge?

**Comment by davies** [comment]
@JoshRosen Considering a case that the used memory is above memory limit (because of broadcasted object), then it will start to spill after the first batch (batch =100), and for every 100 records, it's easy to reach maximum open files limit (1024, by default).

In order to balance the memory usage and batch size (not too large or too small), I thought it's better to adjust batch size up and down (but having a bug). After some experiments, I saw the memory usage still go up and up even we try to shrink the batch size (because of memory fragments).

Finally, I'd like to switch to the most simple approach, assuming the items having similar sizes, always use the first batch size when first spilling.

I agreed that we could inline `_next_limit()`, current it looks similar as others (ExternalXXX), not that bad.

**Comment by JoshRosen** [comment]
Based on this latest round of comments + some offline discussion, this sounds like a reasonable fix for the 1.4.1 release, so I'm going to merge this to `master` and `branch-1.4`.

**Pull request closed without merging**